File: midi-b/midi_b/deid/dateshift.py
"""DICOM date shifting utilities."""

from __future__ import annotations

import hashlib
from datetime import datetime, timedelta

# Tags whose values are dates that should be shifted.
DATE_TAGS: set[str] = {
    "00080012",  # InstanceCreationDate
    "00080020",  # StudyDate
    "00080021",  # SeriesDate
    "00080022",  # AcquisitionDate
    "00080023",  # ContentDate
    "00100030",  # PatientBirthDate
    "00400244",  # PerformedProcedureStepStartDate
}

# Tags whose values are times paired with the above dates — kept unchanged.
TIME_TAGS: set[str] = {
    "00080013",  # InstanceCreationTime
    "00080030",  # StudyTime
    "00080031",  # SeriesTime
    "00080032",  # AcquisitionTime
    "00080033",  # ContentTime
    "00400245",  # PerformedProcedureStepStartTime
}


def shift_dicom_date(date_str: str, offset_days: int) -> str:
    """Shift a DICOM DA value (YYYYMMDD) by *offset_days*.

    Returns the shifted date in YYYYMMDD format.
    If the input is not a valid 8-digit date, returns it unchanged.
    """
    if len(date_str) != 8 or not date_str.isdigit():
        return date_str
    try:
        dt = datetime(int(date_str[:4]), int(date_str[4:6]), int(date_str[6:8]))
    except ValueError:
        return date_str
    shifted = dt + timedelta(days=offset_days)
    return shifted.strftime("%Y%m%d")


def shift_dicom_datetime(dt_str: str, offset_days: int) -> str:
    """Shift a DICOM DT value (YYYYMMDD + optional time/tz suffix).

    Only the date portion (first 8 chars) is shifted; the rest is preserved.
    """
    if len(dt_str) < 8:
        return dt_str
    date_part = shift_dicom_date(dt_str[:8], offset_days)
    return date_part + dt_str[8:]


def generate_date_offset(patient_id: str, salt: str, max_days: int = 365) -> int:
    """Generate a deterministic date-shift offset for a patient.

    Uses SHA-256 of ``salt + patient_id``; first 4 bytes as unsigned 32-bit
    integer mapped to ``[-max_days, +max_days]``.
    """
    data = (salt + patient_id).encode("utf-8")
    digest = hashlib.sha256(data).digest()
    raw = int.from_bytes(digest[:4], byteorder="big")
    span = max_days * 2 + 1
    return (raw % span) - max_days

File: client/src/dicom/dateshift.ts
/**
 * DICOM date shifting utilities for de-identification.
 *
 * Instead of removing or zeroing dates (which destroys temporal relationships),
 * date shifting applies a consistent per-patient day offset to all dates.
 * This preserves relative timing between studies while preventing re-identification.
 */

/** Known DICOM date tags (DA VR) that should be shifted */
export const DATE_TAGS: Set<string> = new Set([
  '00080012', // InstanceCreationDate
  '00080020', // StudyDate
  '00080021', // SeriesDate
  '00080022', // AcquisitionDate
  '00080023', // ContentDate
  '00100030', // PatientBirthDate
  '00400244', // PerformedProcedureStepStartDate
])

/** Known DICOM time tags (TM VR) associated with date tags — kept unchanged */
export const TIME_TAGS: Set<string> = new Set([
  '00080013', // InstanceCreationTime
  '00080030', // StudyTime
  '00080031', // SeriesTime
  '00080032', // AcquisitionTime
  '00080033', // ContentTime
  '00400245', // PerformedProcedureStepStartTime
])

/**
 * Shift a DICOM DA (YYYYMMDD) value by the given number of days.
 * Returns the shifted date in YYYYMMDD format.
 * Returns the original string unchanged if it cannot be parsed.
 */
export function shiftDicomDate(dateStr: string, offsetDays: number): string {
  if (!dateStr || dateStr.length < 8) return dateStr

  const year = parseInt(dateStr.substring(0, 4), 10)
  const month = parseInt(dateStr.substring(4, 6), 10) - 1 // JS months are 0-based
  const day = parseInt(dateStr.substring(6, 8), 10)

  if (isNaN(year) || isNaN(month) || isNaN(day)) return dateStr

  const date = new Date(Date.UTC(year, month, day))
  date.setUTCDate(date.getUTCDate() + offsetDays)

  const y = date.getUTCFullYear().toString().padStart(4, '0')
  const m = (date.getUTCMonth() + 1).toString().padStart(2, '0')
  const d = date.getUTCDate().toString().padStart(2, '0')
  return `${y}${m}${d}`
}

/**
 * Shift a DICOM DT (YYYYMMDDHHMMSS.FFFFFF&ZZXX) value by the given number of days.
 * Only the date portion (first 8 chars) is shifted; the time and timezone are preserved.
 * Returns the original string unchanged if it cannot be parsed.
 */
export function shiftDicomDateTime(dtStr: string, offsetDays: number): string {
  if (!dtStr || dtStr.length < 8) return dtStr

  const datePart = dtStr.substring(0, 8)
  const rest = dtStr.substring(8)
  const shifted = shiftDicomDate(datePart, offsetDays)

  return shifted + rest
}

/**
 * Generate a deterministic day offset from a patient ID and salt.
 *
 * Uses SHA-256 of (salt + patientId), takes the first 4 bytes as an unsigned
 * integer, and maps it to the range [-maxDays, +maxDays].
 *
 * The same patient ID and salt always produce the same offset, ensuring
 * consistency across all files for the same patient.
 *
 * NOTE: This function is async because it uses the Web Crypto API (crypto.subtle).
 */
export async function generateDateOffset(
  patientId: string,
  salt: string,
  maxDays: number = 365
): Promise<number> {
  const encoder = new TextEncoder()
  const data = encoder.encode(salt + patientId)
  const hashBuffer = await crypto.subtle.digest('SHA-256', data)
  const view = new DataView(hashBuffer)

  // Read first 4 bytes as unsigned 32-bit integer
  const raw = view.getUint32(0)

  // Map to [-maxDays, +maxDays]
  const range = maxDays * 2 + 1
  const offset = (raw % range) - maxDays

  return offset
}
